openupgrade_scripts/scripts/survey/14.0.3.2/pre-migration.py

```diff
@@ -0,0 +1,242 @@
+        "state",
+        [("skip", "in_progress")],
+        table="survey_user_input",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE survey_survey
+        ADD COLUMN session_state varchar""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE survey_survey
+        ADD COLUMN session_start_time timestamp without time zone""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE survey_survey ss
+        SET session_state = CASE WHEN sui.state = 'new' THEN 'ready'
+            WHEN sui.state = 'in_progress' THEN 'in_progress'
+            ELSE NULL END, session_start_time = CASE
+            WHEN sui.state IN ('new', 'in_progress') THEN ss.write_date
+            ELSE NULL END
+        FROM survey_user_input sui
+        WHERE sui.survey_id = ss.id AND ss.state != 'draft'""",
+    )
+
+
+def fast_fill_survey_input_some_values(env):
+    # some new computes depend of this is_session_answer field
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE survey_user_input
+        ADD COLUMN is_session_answer boolean""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE survey_user_input
+        ADD COLUMN nickname varchar""",
+    )
+    # is_session_answer
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE survey_user_input sui
+        SET is_session_answer = TRUE, start_datetime = sui.write_date
+        FROM survey_survey ss
+        WHERE sui.survey_id = ss.id AND ss.session_state IN ('ready', 'in_progress')""",
+    )
+    # nickname
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE survey_user_input sui
+        SET nickname = rp.name
+        FROM res_partner rp
+        WHERE sui.partner_id = rp.id""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE survey_user_input sui
+        SET nickname = sui.email
+        WHERE sui.nickname IS NULL""",
+    )
+
+
+def fill_survey_user_input_line_answer_is_correct(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE survey_user_input_line
+        ADD COLUMN answer_is_correct boolean""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE survey_user_input_line suil
+        SET answer_is_correct = sqa.is_correct
+        FROM survey_question_answer sqa, survey_question sq
+        WHERE suil.suggested_answer_id = sqa.id AND suil.question_id = sq.id
+            AND sq.question_type IN ('simple_choice', 'multiple_choice')""",
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.copy_columns(env.cr, _column_copies)
+    openupgrade.rename_fields(env, _field_renames)
+    openupgrade.rename_models(env.cr, _model_renames)
+    openupgrade.rename_tables(env.cr, _table_renames)
+    openupgrade.rename_xmlids(env.cr, _xmlid_renames)
```

Comment:
Should the new field `session_code` be filled here?

This field has default `_get_default_session_code()` but it would generate the same codes for all rows and break the _sql_constraints `('session_code_unique', 'unique(session_code)', 'Session code should be unique'),`

Draft idea is something like this
```python
def _fill_survey_session_code(env):
    openupgrade.logged_query(
        env.cr,
        "ALTER TABLE survey_survey ADD COLUMN session_code character varying;"
        )
    env.cr.execute(
        env.cr,
        "SELECT id FROM survey_survey ORDER BY id ASC;"
        )
    rows = env.cr.dictfetchall()
    start_code = 1000
    for row in rows:
        openupgrade.logged_query(
            env.cr,
            """
            UPDATE survey_survey SET session_code = %s WHERE id=%s;
            """,
            (start_code, row['id'])
            )
        start_code += 1
```

Reply:
filled